ascon-aead/src/asconcore.rs

```diff
@@ -0,0 +1,379 @@
+pub(crate) trait InternalKey<KS: ArrayLength<u8>>:
+    Sized + Clone + for<'a> From<&'a GenericArray<u8, KS>>
+{
+    /// Return K0.
+    fn get_k0(&self) -> u64;
+    /// Return K1.
+    fn get_k1(&self) -> u64;
+    /// Return K2.
+    fn get_k2(&self) -> u64;
+}
+
+#[derive(Clone)]
+#[cfg_attr(feature = "zeroize", derive(zeroize::Zeroize, zeroize::ZeroizeOnDrop))]
+pub(crate) struct InternalKey16(u64, u64);
+
+impl InternalKey<U16> for InternalKey16 {
+    #[inline(always)]
+    fn get_k0(&self) -> u64 {
+        0
+    }
+
+    #[inline(always)]
+    fn get_k1(&self) -> u64 {
+        self.0
+    }
+
+    #[inline(always)]
+    fn get_k2(&self) -> u64 {
+        self.1
+    }
+}
+
+impl From<&GenericArray<u8, U16>> for InternalKey16 {
+    fn from(key: &GenericArray<u8, U16>) -> Self {
+        Self(u64_from_be_bytes(&key[..8]), u64_from_be_bytes(&key[8..]))
+    }
+}
+
+#[derive(Clone)]
+#[cfg_attr(feature = "zeroize", derive(zeroize::Zeroize, zeroize::ZeroizeOnDrop))]
+pub(crate) struct InternalKey20(u64, u64, u32);
+
+impl InternalKey<U20> for InternalKey20 {
+    #[inline(always)]
+    fn get_k0(&self) -> u64 {
+        self.2 as u64
+    }
+
+    #[inline(always)]
+    fn get_k1(&self) -> u64 {
+        self.0
+    }
+
+    #[inline(always)]
+    fn get_k2(&self) -> u64 {
+        self.1
+    }
+}
+
+impl From<&GenericArray<u8, U20>> for InternalKey20 {
+    fn from(key: &GenericArray<u8, U20>) -> Self {
+        Self(
+            u64_from_be_bytes(&key[4..12]),
+            u64_from_be_bytes(&key[12..]),
+            u32_from_be_bytes(&key[..4]),
+        )
+    }
+}
+
+/// Parameters of an Ascon instance
+pub(crate) trait Parameters {
+    /// Size of the secret key
+    ///
+    /// For internal use-only.
+    type KeySize: ArrayLength<u8>;
+    /// Internal storage for secret keys
+    ///
+    /// For internal use-only.
+    type InternalKey: InternalKey<Self::KeySize>;
+
+    /// Number of bytes to process per round
+    const COUNT: usize;
+    /// Initialization vector used to initialize Ascon's state
+    ///
+    /// For internal use-only
+    const IV: u64;
+}
+
+/// Parameters for Ascon-128
+pub(crate) struct Parameters128;
+
+impl Parameters for Parameters128 {
+    type KeySize = U16;
+    type InternalKey = InternalKey16;
+
+    const COUNT: usize = 8;
+    const IV: u64 = 0x80400c0600000000;
+}
+
+/// Parameters for Ascon-128a
+pub(crate) struct Parameters128a;
+
+impl Parameters for Parameters128a {
+    type KeySize = U16;
+    type InternalKey = InternalKey16;
+
+    const COUNT: usize = 16;
+    const IV: u64 = 0x80800c0800000000;
+}
+
+/// Parameters for Ascon-80pq
+pub(crate) struct Parameters80pq;
+
+impl Parameters for Parameters80pq {
+    type KeySize = U20;
+    type InternalKey = InternalKey20;
+
+    const COUNT: usize = 8;
+    const IV: u64 = 0xa0400c0600000000;
+}
+
+/// Core implementation of Ascon for one encryption/decryption operation
+pub(crate) struct AEADCore<'a, P: Parameters> {
```

Comment:
I think this should be `AeadCore` per https://rust-lang.github.io/api-guidelines/naming.html

> In UpperCamelCase, acronyms and contractions of compound words count as one word: use `Uuid` rather than `UUID`

Reply:
Renamed to `AsconCore`.